### DISCUSSION TITLE
Extensions: how to save extension's config?

### DISCUSSION THREAD

**Discussion by rjzak** [NationalSecurityAgency/ghidra#5705] [Q&A]
I'm working on an extension which communicates to a server for additional analytics. I need to be able to save the server's config, so the user doesn't have to provide it each time. I have some functions which seem to be what's needed to store/retrieve this data, but I don't see any evidence that they're being called.

* What am I missing?
* How can I have the data loaded from config at startup?

I'm hoping there's a way for the extension to store it's data as part of Ghidra's configuration, wherever that is. I'm hoping I don't have to read/write a config file independently, but I suppose that's not the end of the world. I'd like to do it "the Ghidra way" if possible.

```java
@Override
public void readConfigState(SaveState saveState) {
    Msg.debug(this, "readConfigState() called");
    client = new Client(saveState.getString("SERVER_ADDR", SERVER_ADDRESS_DEFAULT));
    Msg.debug(this, "Client object updated, got server address: " + client.url);
}

@Override
public void writeConfigState(SaveState saveState) {
    Msg.debug(this, "writeConfigState() called");
    saveState.putString("SERVER_ADDR", client.url);
}
```

Thank you!

**Comment by dev747368** [comment]
It does seem strange that your plugin's (read|write)ConfigState() isn't getting called.  Seems like some debugging is in order for how your plugin is getting constructed and initialized.

However, for user-visible configuration items like a host address, it might be better to save them as an option using the PluginTool's getOptions() method.  This will allow your config value to be present in the gui config screen (Edit | Tool Options)
There are several examples of other plugins calling ```tool.getOptions(THEIR_PLUGIN_CONFIG_NAME)```.  (you have to register a listener to get notified of when the options are changed by the user)  ProgramTreePlugin.java:125 is one location that you might use as an example, but that is not the only way to do it.

**Comment by dragonmacher** [comment]
A couple things to point out:

- `readConfigState()` will only get called if you have written something in a previous call to `writeConfigState()`
- `writeConfigState()` should be getting called every time the tool is saved, such as when closing the tool

Also, these methods will not be called if you have not 'configured' your extension to be part of the tool.  To do this, from the menu, press `File -> Configure` to show all plugins known to the tool.    From that dialog, find your extension's plugin(s) and make sure they are installed in the tool (with the checkbox being checked).   After that, you may need to close and reopen the tool for the methods to start getting called.

I verified these methods get called when installed in the tool.
